Several native integration tests using JDK provided service provider implementations from modules fail with latest Graal master

### Describe the bug

We see a couple of quarkus integration tests failing with latest GraalVM master:

### Elytron Security LDAP

```
2022-12-21 03:08:50,246 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /api/requiresAdminRole failed, error id: 03e5088d-47fb-4128-b7f5-b1024c6f372d-1: java.lang.RuntimeException: org.wildfly.security.auth.server.RealmUnavailableException: ELY01125: Ldap-backed realm failed to obtain context
	at io.quarkus.elytron.security.runtime.ElytronPasswordIdentityProvider$1.get(ElytronPasswordIdentityProvider.java:62)
	at io.quarkus.elytron.security.runtime.ElytronPasswordIdentityProvider$1.get(ElytronPasswordIdentityProvider.java:43)
	at io.quarkus.security.runtime.QuarkusIdentityProviderManagerImpl$1$1$1$1.run(QuarkusIdentityProviderManagerImpl.java:58)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@17.0.6-beta/java.lang.Thread.run(Thread.java:833)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:810)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211)
Caused by: org.wildfly.security.auth.server.RealmUnavailableException: ELY01125: Ldap-backed realm failed to obtain context
	at org.wildfly.security.auth.realm.ldap.LdapSecurityRealm.obtainContext(LdapSecurityRealm.java:223)
	at org.wildfly.security.auth.realm.ldap.LdapSecurityRealm.access$600(LdapSecurityRealm.java:104)
	at org.wildfly.security.auth.realm.ldap.LdapSecurityRealm$LdapRealmIdentity.verifyEvidence(LdapSecurityRealm.java:607)
	at org.wildfly.security.auth.server.ServerAuthenticationContext$NameAssignedState.verifyEvidence(ServerAuthenticationContext.java:2078)
	at org.wildfly.security.auth.server.ServerAuthenticationContext.verifyEvidence(ServerAuthenticationContext.java:767)
	at org.wildfly.security.auth.server.SecurityDomain.authenticate(SecurityDomain.java:315)
	at org.wildfly.security.auth.server.SecurityDomain.authenticate(SecurityDomain.java:289)
	at io.quarkus.elytron.security.runtime.ElytronPasswordIdentityProvider$1.get(ElytronPasswordIdentityProvider.java:48)
	... 11 more
Caused by: javax.naming.NoInitialContextException: Cannot load initial context factory 'com.sun.jndi.ldap.LdapCtxFactory' [Root exception is java.util.ServiceConfigurationError: javax.naming.spi.InitialContextFactory: Provider com.sun.jndi.dns.DnsContextFactory not found]
	at java.naming@17.0.6-beta/javax.naming.spi.NamingManager.getFactory(NamingManager.java:748)
	at java.naming@17.0.6-beta/javax.naming.spi.NamingManager.lambda$getInitialContext$1(NamingManager.java:724)
	at java.base@17.0.6-beta/jdk.internal.loader.AbstractClassLoaderValue$Memoizer.get(AbstractClassLoaderValue.java:329)
	at java.base@17.0.6-beta/jdk.internal.loader.AbstractClassLoaderValue.computeIfAbsent(AbstractClassLoaderValue.java:205)
	at java.naming@17.0.6-beta/javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:724)
	at java.naming@17.0.6-beta/javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:305)
	at java.naming@17.0.6-beta/javax.naming.InitialContext.init(InitialContext.java:236)
	at java.naming@17.0.6-beta/javax.naming.ldap.InitialLdapContext.<init>(InitialLdapContext.java:154)
	at io.quarkus.elytron.security.ldap.QuarkusDirContextFactory.createDirContext(QuarkusDirContextFactory.java:111)
	at io.quarkus.elytron.security.ldap.QuarkusDirContextFactory.obtainDirContext(QuarkusDirContextFactory.java:51)
	at io.quarkus.elytron.security.ldap.LdapRecorder.lambda$createDirContextSupplier$0(LdapRecorder.java:60)
	at org.wildfly.security.auth.realm.ldap.LdapSecurityRealm.obtainContext(LdapSecurityRealm.java:211)
	... 18 more
Caused by: java.util.ServiceConfigurationError: javax.naming.spi.InitialContextFactory: Provider com.sun.jndi.dns.DnsContextFactory not found
	at java.base@17.0.6-beta/java.util.ServiceLoader.fail(ServiceLoader.java:593)
	at java.base@17.0.6-beta/java.util.ServiceLoader.loadProvider(ServiceLoader.java:875)
	at java.base@17.0.6-beta/java.util.ServiceLoader$ModuleServicesLookupIterator.hasNext(ServiceLoader.java:1084)
	at java.base@17.0.6-beta/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base@17.0.6-beta/java.util.ServiceLoader$ProviderSpliterator.tryAdvance(ServiceLoader.java:1491)
	at java.base@17.0.6-beta/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base@17.0.6-beta/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base@17.0.6-beta/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
	at java.naming@17.0.6-beta/javax.naming.spi.NamingManager.getFactory(NamingManager.java:744)
	... 29 more
```

### Kafka SASL
```
2022-12-21 02:34:25,904 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /sasl failed, error id: 9e427950-b975-4485-aa8a-6679810717ff-1: org.jboss.resteasy.spi.UnhandledException: org.apache.kafka.common.KafkaException: Failed to construct kafka consumer
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:368)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:147)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@17.0.6-beta/java.lang.Thread.run(Thread.java:833)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:810)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211)
Caused by: org.apache.kafka.common.KafkaException: Failed to construct kafka consumer
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:830)
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:666)
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:647)
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:627)
	at io.quarkus.it.kafka.sasl.SaslKafkaEndpoint.createConsumer(SaslKafkaEndpoint.java:66)
	at io.quarkus.it.kafka.sasl.SaslKafkaEndpoint.create(SaslKafkaEndpoint.java:35)
	at io.quarkus.it.kafka.sasl.SaslKafkaEndpoint_Bean.create(Unknown Source)
	at io.quarkus.it.kafka.sasl.SaslKafkaEndpoint_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:113)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:37)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:34)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:34)
	at io.quarkus.it.kafka.sasl.SaslKafkaEndpoint_Bean.get(Unknown Source)
	at io.quarkus.it.kafka.sasl.SaslKafkaEndpoint_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:476)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:489)
	at io.quarkus.arc.impl.ArcContainerImpl$2.get(ArcContainerImpl.java:279)
	at io.quarkus.arc.impl.ArcContainerImpl$2.get(ArcContainerImpl.java:276)
	at io.quarkus.resteasy.common.runtime.QuarkusConstructorInjector.construct(QuarkusConstructorInjector.java:52)
	at org.jboss.resteasy.plugins.server.resourcefactory.POJOResourceFactory.createResource(POJOResourceFactory.java:71)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:401)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	... 17 more
Caused by: org.apache.kafka.common.KafkaException: java.util.ServiceConfigurationError: javax.security.auth.spi.LoginModule: Provider com.sun.security.auth.module.Krb5LoginModule not found
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:184)
	at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:192)
	at org.apache.kafka.common.network.ChannelBuilders.clientChannelBuilder(ChannelBuilders.java:81)
	at org.apache.kafka.clients.ClientUtils.createChannelBuilder(ClientUtils.java:105)
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:738)
	... 41 more
Caused by: java.util.ServiceConfigurationError: javax.security.auth.spi.LoginModule: Provider com.sun.security.auth.module.Krb5LoginModule not found
	at java.base@17.0.6-beta/java.util.ServiceLoader.fail(ServiceLoader.java:593)
	at java.base@17.0.6-beta/java.util.ServiceLoader.loadProvider(ServiceLoader.java:875)
	at java.base@17.0.6-beta/java.util.ServiceLoader$ModuleServicesLookupIterator.hasNext(ServiceLoader.java:1084)
	at java.base@17.0.6-beta/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base@17.0.6-beta/java.util.ServiceLoader$ProviderSpliterator.tryAdvance(ServiceLoader.java:1491)
	at java.base@17.0.6-beta/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base@17.0.6-beta/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base@17.0.6-beta/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext.invoke(LoginContext.java:713)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext$4.run(LoginContext.java:679)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext$4.run(LoginContext.java:677)
	at java.base@17.0.6-beta/java.security.AccessController.executePrivileged(AccessController.java:147)
	at java.base@17.0.6-beta/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:677)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext.login(LoginContext.java:587)
	at org.apache.kafka.common.security.authenticator.AbstractLogin.login(AbstractLogin.java:60)
	at org.apache.kafka.common.security.authenticator.LoginManager.<init>(LoginManager.java:62)
	at org.apache.kafka.common.security.authenticator.LoginManager.acquireLoginManager(LoginManager.java:105)
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:170)
	... 45 more
```

### Kafka OAUTH with Keycloak 
```
2022-12-21 02:41:14,393 ERROR [io.sma.rea.mes.provider] (main) SRMSG00230: Unable to create the publisher or subscriber during initialization: org.apache.kafka.common.KafkaException: Failed to construct kafka consumer
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:830)
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:666)
	at io.smallrye.reactive.messaging.kafka.impl.ReactiveKafkaConsumer.lambda$new$0(ReactiveKafkaConsumer.java:92)
	at java.base@17.0.6-beta/java.util.concurrent.atomic.AtomicReference.updateAndGet(AtomicReference.java:210)
	at io.smallrye.reactive.messaging.kafka.impl.ReactiveKafkaConsumer.lambda$new$1(ReactiveKafkaConsumer.java:88)
	at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:28)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniMemoizeOp.subscribe(UniMemoizeOp.java:84)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at java.base@17.0.6-beta/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base@17.0.6-beta/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base@17.0.6-beta/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base@17.0.6-beta/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base@17.0.6-beta/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base@17.0.6-beta/java.lang.Thread.run(Thread.java:833)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:810)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211)
Caused by: org.apache.kafka.common.KafkaException: java.util.ServiceConfigurationError: javax.security.auth.spi.LoginModule: Provider com.sun.security.auth.module.Krb5LoginModule not found
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:184)
	at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:192)
	at org.apache.kafka.common.network.ChannelBuilders.clientChannelBuilder(ChannelBuilders.java:81)
	at org.apache.kafka.clients.ClientUtils.createChannelBuilder(ClientUtils.java:105)
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:738)
	... 18 more
Caused by: java.util.ServiceConfigurationError: javax.security.auth.spi.LoginModule: Provider com.sun.security.auth.module.Krb5LoginModule not found
	at java.base@17.0.6-beta/java.util.ServiceLoader.fail(ServiceLoader.java:593)
	at java.base@17.0.6-beta/java.util.ServiceLoader.loadProvider(ServiceLoader.java:875)
	at java.base@17.0.6-beta/java.util.ServiceLoader$ModuleServicesLookupIterator.hasNext(ServiceLoader.java:1084)
	at java.base@17.0.6-beta/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base@17.0.6-beta/java.util.ServiceLoader$ProviderSpliterator.tryAdvance(ServiceLoader.java:1491)
	at java.base@17.0.6-beta/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base@17.0.6-beta/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base@17.0.6-beta/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base@17.0.6-beta/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext.invoke(LoginContext.java:713)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext$4.run(LoginContext.java:679)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext$4.run(LoginContext.java:677)
	at java.base@17.0.6-beta/java.security.AccessController.executePrivileged(AccessController.java:147)
	at java.base@17.0.6-beta/java.security.AccessController.doPrivileged(AccessController.java:712)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:677)
	at java.base@17.0.6-beta/javax.security.auth.login.LoginContext.login(LoginContext.java:587)
	at org.apache.kafka.common.security.oauthbearer.internals.expiring.ExpiringCredentialRefreshingLogin.login(ExpiringCredentialRefreshingLogin.java:204)
	at org.apache.kafka.common.security.oauthbearer.internals.OAuthBearerRefreshingLogin.login(OAuthBearerRefreshingLogin.java:150)
	at org.apache.kafka.common.security.authenticator.LoginManager.<init>(LoginManager.java:62)
	at org.apache.kafka.common.security.authenticator.LoginManager.acquireLoginManager(LoginManager.java:105)
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:170)
	... 22 more
```

### JPA Oracle

This is a bit more cryptic but it looks like it boils down to LDAP again
```
Response body doesn't match expectation.
Expected: is "OK"
  Actual: An error occurred while attempting ldap operations
```

### Expected behavior

Tests pass.

### Actual behavior

Tests fail.

### How to Reproduce?

Run those quarkus integration tests in native mode with latest graalvm master.

### Output of `uname -a` or `ver`

Linux

### Output of `java -version`

JDK 17

### GraalVM version (if different from Java)

Graal VM master (rev 16c4eaaf32a525ac327bc596927d96ca70f0e16a)

### Quarkus version or git rev

`main` branch

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

This seems to be caused by: https://github.com/oracle/graal/pull/5601

Confirmed with GHA runs prior and after the change.

Revision just before the change (good): 5f0ff219a77797d08a0cea57ea773b513c30237f
https://github.com/graalvm/mandrel/actions/runs/3749623804

Last commit of this change (bad): 56877161d7f964099112b9a251579391412f3413
https://github.com/graalvm/mandrel/actions/runs/3749641198

--- a/extensions/elytron-security-ldap/deployment/src/main/java/io/quarkus/elytron/security/ldap/deployment/ElytronSecurityLdapProcessor.java
+++ b/extensions/elytron-security-ldap/deployment/src/main/java/io/quarkus/elytron/security/ldap/deployment/ElytronSecurityLdapProcessor.java
@@ -61,7 +61,11 @@ class ElytronSecurityLdapProcessor {
     }
 
     @BuildStep
-    ReflectiveClassBuildItem enableReflection() {
-        return new ReflectiveClassBuildItem(true, true, QuarkusDirContextFactory.INITIAL_CONTEXT_FACTORY);
+    void registerForReflection(BuildProducer<ReflectiveClassBuildItem> reflection) {
+        // All JDK provided InitialContextFactory impls via the module descriptors:
+        // com.sun.jndi.ldap.LdapCtxFactory, com.sun.jndi.dns.DnsContextFactory and com.sun.jndi.rmi.registry.RegistryContextFactory
+        reflection.produce(new ReflectiveClassBuildItem(true, true, QuarkusDirContextFactory.INITIAL_CONTEXT_FACTORY));
+        reflection.produce(new ReflectiveClassBuildItem(false, false, "com.sun.jndi.dns.DnsContextFactory"));
+        reflection.produce(new ReflectiveClassBuildItem(false, false, "com.sun.jndi.rmi.registry.RegistryContextFactory"));
     }
 }

--- a/extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleNativeImage.java
+++ b/extensions/jdbc/jdbc-oracle/deployment/src/main/java/io/quarkus/jdbc/oracle/deployment/OracleNativeImage.java
@@ -25,7 +25,14 @@ public final class OracleNativeImage {
         reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, driverName));
 
         // for ldap style jdbc urls. e.g. jdbc:oracle:thin:@ldap://oid:5000/mydb1,cn=OracleContext,dc=myco,dc=com
+        //
+        // Note that all JDK provided InitialContextFactory impls from the JDK registered via module descriptors
+        // available at build time need to be reflectively accessible via ServiceLoader for runtime consistency.
+        // These are:
+        // com.sun.jndi.ldap.LdapCtxFactory, com.sun.jndi.dns.DnsContextFactory and com.sun.jndi.rmi.registry.RegistryContextFactory
         reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.jndi.ldap.LdapCtxFactory"));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.jndi.dns.DnsContextFactory"));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.jndi.rmi.registry.RegistryContextFactory"));
     }
 
     @BuildStep

--- a/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
+++ b/extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
@@ -433,9 +433,23 @@ public class KafkaProcessor {
         for (ClassInfo loginModule : index.getIndex().getAllKnownImplementors(LOGIN_MODULE)) {
             reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, loginModule.name().toString()));
         }
+        // Kafka oauth login internally iterates over all ServiceLoader available LoginModule's
+        registerJDKLoginModules(reflectiveClass);
         for (ClassInfo authenticateCallbackHandler : index.getIndex().getAllKnownImplementors(AUTHENTICATE_CALLBACK_HANDLER)) {
             reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, authenticateCallbackHandler.name().toString()));
         }
+    }
+
+    private void registerJDKLoginModules(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
+        // jdk.security.auth module provided LoginModule's
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.security.auth.module.Krb5LoginModule"));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.security.auth.module.UnixLoginModule"));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.security.auth.module.JndiLoginModule"));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.security.auth.module.KeyStoreLoginModule"));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.security.auth.module.LdapLoginModule"));
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.security.auth.module.NTLoginModule"));
+        // java.management module provided LoginModule's
+        reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, "com.sun.jmx.remote.security.FileLoginModule"));
     }
 
     private static void collectImplementors(Set<DotName> set, CombinedIndexBuildItem indexBuildItem, Class<?> cls) {

